### PR TITLE
Guard against exceptions in to_s when cleaning objects

### PR DESCRIPTION
**Pull request by imjoehaines**
## Goal

When sending notifications we run through a "cleaning" stage which does things like filter out any keys that we shouldn't send in the error report (e.g. passwords). This calls `to_s` in a couple of different places, which can cause issues as objects are free to implement any logic and can therefore throw exceptions or infinitely recurse, which means we end up crashing while trying to cleanup the object

In this PR we'll now recover from exceptions and stack overflows by replacing the values with one of our existing replacement strings:

1. `[RECURSION]` if we get a `SystemStackError` when stringifying a plain object
1. `[FILTERED]` if we get an exception when attempting to filter keys in a hash map — if an exception happens we can't tell if we should filter the key, so we have to assume we need to

This means the error report will have less information that intended, but this should be rare and if it does happen then the alternative is no report at all

### Changed

- Rescue `SystemStackError` as well as other exceptions when calling `to_s` on an object
- Rescue `StandardError` and `SystemStackError` when calling `to_s` on a key when filtering a hash map

## Tests

Tests have been added for:
- hash keys that throw during `to_s`
- hash values that throw during `to_s`
- custom objects that infinitely recurse during `to_s`

### Linked issues

Related to #577

## Review

For the submitter, initial self-review:

- [x] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [x] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: lib/bugsnag/cleaner.rb [hunk 1]
```diff
@@ -43,7 +43,13 @@ def traverse_object(obj, seen, scope)
       when String
         clean_string(obj)
       else
-        str = obj.to_s rescue RAISED
+        # guard against objects that raise or blow the stack when stringified
+        begin
+          str = obj.to_s rescue RAISED
```
Comment (tomlongridge) [inline lib/bugsnag/cleaner.rb:48]: Does the `rescue RAISED` not catch all errors already?

Comment (imjoehaines) [reply lib/bugsnag/cleaner.rb:48]: It catches all errors that inherit from `StandardError`, but `SystemStackError` doesn't so it won't be caught here without the separate `rescue`

I can make another `rescue` block for `StandardError` explicitly if it's confusing; something like:

```ruby
begin
  str = obj.to_s
rescue StandardError
  str = RAISED
rescue SystemStackError
  str = RECURSION
end
```

Comment (tomlongridge) [reply lib/bugsnag/cleaner.rb:48]: Cool - yes, that would help (me!)

Comment (imjoehaines) [reply lib/bugsnag/cleaner.rb]: Made the change in a1ad37e :)
